qmp/libvirtgo_linux.go

```diff
@@ -0,0 +1,327 @@
+	}
+
+	mon.domain = &domain
+	mon.virConn = &virConn
+	mon.eventsChan = make(chan Event)
+	mon.doneChan = make(chan bool)
+
+	return err
+}
+
+var newVirConnectionInternal = func(
+	uri string) (libvirt.VirConnection, error) {
+	return libvirt.NewVirConnection(uri)
+}
+
+// Disconnect tears down open QMP socket connections.
+func (mon *libvirtGoMonitorLinux) Disconnect() error {
+	var err error
+	if mon.domain != nil {
+		mon.domain.Free()
+		mon.domain = nil
+	}
+
+	if mon.virConn != nil {
+		_, err = closeConnectionInternal(mon.virConn)
+		mon.virConn = nil
+		close(mon.doneChan)   // stop listenning to events
+		close(mon.eventsChan) // stop sending events to clients
+	}
+
+	return err
+}
+
+var closeConnectionInternal = func(
+	virConn *libvirt.VirConnection) (int, error) {
+	return virConn.CloseConnection()
+}
+
+// Run executes the given QAPI command against a domain's QEMU instance.
+// For a list of available QAPI commands, see:
+//	http://git.qemu.org/?p=qemu.git;a=blob;f=qapi-schema.json;hb=HEAD
+func (mon *libvirtGoMonitorLinux) Run(cmd []byte) ([]byte, error) {
+	if mon.virConn == nil {
+		return nil, errors.New("Run() need a established connection to proceed.")
+	}
+
+	result, err := qemuMonitorCommandInternal(mon.domain, string(cmd))
+	if err != nil {
+		return nil, err
+	}
+
+	return []byte(result), nil
+}
+
+var qemuMonitorCommandInternal = func(
+	domain *libvirt.VirDomain, cmd string) (string, error) {
+	return domain.QemuMonitorCommand(
+		libvirt.VIR_DOMAIN_QEMU_MONITOR_COMMAND_DEFAULT,
+		cmd)
+}
+
+// Events streams QEMU QMP Events.
+// If a problem is encountered setting up the event monitor connection
+// an error will be returned. Errors encountered during streaming will
+// cause the returned event channel to be closed.
+func (mon *libvirtGoMonitorLinux) Events() (<-chan Event, error) {
+
+	if mon.virConn == nil {
+		return nil,
+			errors.New("Events() need a established connection to proceed.")
+	}
+
+	callbackID, err := domainEventRegister(mon)
+	if err != nil {
+		return nil, err
+	}
+
+	go pollEventsLoop(mon.doneChan)
+
+	go domainEventDeregister(mon, callbackID)
+
+	return mon.eventsChan, nil
+}
+
+var lookupDomainByName = func(
+	virConn *libvirt.VirConnection,
+	domainName string) (libvirt.VirDomain, error) {
+	return virConn.LookupDomainByName(domainName)
+}
+
+// eventRegisterDefaultImpl registers a default event implementation.
+func eventRegisterDefaultImpl() {
+	eventRegisterID := eventRegisterDefaultImplInternal()
+	if eventRegisterID == -1 {
+		log.Printf(
+			"EventRegisterDefaultImpl returned an unexpected value %d\n",
+			eventRegisterID)
+		//TODO: panic or what???
+	}
+}
+
+var eventRegisterDefaultImplInternal = func() int {
+	return libvirt.EventRegisterDefaultImpl()
+}
+
+// domainEventRegister register with libvirt to receive events for the
+// specified domain.
+func domainEventRegister(mon *libvirtGoMonitorLinux) (int, error) {
+	callback :=
+		libvirt.DomainEventCallback(newEventCallback(mon))
```

Comment:
Move this up on the same line.  Only break up excessively long or unreadable lines.